plugin/src/main/java/org/opennms/timeseries/cortex/CortexTSS.java

```diff
@@ -213,6 +235,54 @@ public void store(final List<Sample> samples, String clientID) throws StorageExc
         });
     }
 
+    private void persistExternalTags(final Sample s) {
+        // save external tags on a separate database
+        String key = s.getMetric().getKey();
+        var externalTags = externalTagsCache.getIfPresent(key);
+        boolean needUpsert = false;
+
+        JSONObject jsonMetrics = null;
+        JSONObject jsonNewMetric = new JSONObject();
+
+        if (config.getExternalCacheSize() > 0 && externalTags != null) {
+            jsonMetrics = new JSONObject(externalTags);
+            for (Tag tag : s.getMetric().getExternalTags()) {
+                if (!jsonMetrics.has(tag.getKey())) {
+                    jsonMetrics.put(tag.getKey(), tag.getValue());
+                    needUpsert = true;
+                }
+            }
+            if (needUpsert) {
+                kvStore.putAsync(key, jsonMetrics.toString(), CORTEX_TSS);
+                extTagsModified.mark();
+            }
+            extTagsCacheUsed.mark();
+        } else {
+            var externalMetricFromDb = kvStore.get(s.getMetric().getKey(), CORTEX_TSS);
+            if (externalMetricFromDb.isPresent()) {
+                jsonMetrics = new JSONObject(externalMetricFromDb.get().toString());
+                for (Tag tag : s.getMetric().getExternalTags()) {
+                    if (!jsonMetrics.has(tag.getKey())) {
+                        jsonMetrics.put(tag.getKey(), tag.getValue());
+                        needUpsert = true;
+                    }
+                }
+                if (needUpsert) {
+                    kvStore.putAsync(key, jsonMetrics.toString(), CORTEX_TSS);
```

Comment:
I am not sure but don't we need to also put updated value in cache?
externalTagsCache.put(key, jsonNewMetric.toString());